services/notifier/notifier.go

```diff
@@ -0,0 +1,588 @@
+	return nil
+}
+
+func (n *Notifier) setupTables() error {
+	m := &migrator.Migrator{
+		Handle:                     n.db.DB,
+		MigrationsTable:            "pg_notifier_queue_migrations",
+		ShouldForceSetLowerVersion: n.config.shouldForceSetLowerVersion,
+	}
+
+	operation := func() error {
+		return m.Migrate("pg_notifier_queue")
+	}
+
+	backoffWithMaxRetry := backoff.WithMaxRetries(backoff.NewExponentialBackOff(), 3)
+
+	err := backoff.RetryNotify(operation, backoffWithMaxRetry, func(err error, t time.Duration) {
+		n.logger.Warnf("retrying warehouse database migration in %s: %v", t, err)
+	})
+	if err != nil {
+		return fmt.Errorf("could not migrate pg_notifier_queue: %w", err)
+	}
+	return nil
+}
+
+// ClearJobs deletes all jobs for the current workspace.
+func (n *Notifier) ClearJobs(ctx context.Context) error {
+	if n.workspaceIdentifier == "" {
+		return nil
+	}
+
+	n.logger.Infof("Deleting all jobs for workspace: %s", n.workspaceIdentifier)
+
+	err := n.repo.resetForWorkspace(ctx, n.workspaceIdentifier)
+	if err != nil {
+		return fmt.Errorf("could not reset notifier for workspace: %s: %w", n.workspaceIdentifier, err)
+	}
+	return nil
+}
+
+func (n *Notifier) GetDBHandle() *sql.DB {
+	return n.db.DB
+}
+
+// Publish inserts the payloads into the database and returns a channel of type PublishResponse
+func (n *Notifier) Publish(
+	ctx context.Context,
+	publishRequest *PublishRequest,
+) (<-chan *PublishResponse, error) {
+	publishStartTime := n.now()
+
+	batchID := n.batchIDGenerator().String()
+
+	if err := n.repo.insert(ctx, publishRequest, n.workspaceIdentifier, batchID); err != nil {
+		return nil, fmt.Errorf("inserting jobs: %w", err)
+	}
+
+	n.logger.Infof("Inserted %d records into %s for batch: %s", len(publishRequest.Payloads), queueName, batchID)
+
+	n.stats.insertRecords.Count(len(publishRequest.Payloads))
+
+	defer func() {
+		n.stats.publishTime.Since(publishStartTime)
+		n.stats.publish.Increment()
+	}()
+
+	return n.trackBatch(ctx, batchID), nil
+}
+
+// trackBatch tracks the batch and returns a channel of type PublishResponse
+func (n *Notifier) trackBatch(
+	ctx context.Context,
+	batchID string,
+) <-chan *PublishResponse {
+	publishResCh := make(chan *PublishResponse)
+
+	n.background.group.Go(func() error {
+		defer close(publishResCh)
+
+		onUpdate := func(response *PublishResponse) {
+			select {
+			case <-ctx.Done():
+				return
+			case <-n.background.groupCtx.Done():
+				return
+			case publishResCh <- response:
+			}
+		}
+
+		for {
+			select {
+			case <-ctx.Done():
+				return nil
+			case <-n.background.groupCtx.Done():
+				return nil
+			case <-time.After(n.config.trackBatchInterval):
+			}
+
+			count, err := n.repo.pendingByBatchID(ctx, batchID)
+			if err != nil {
+				onUpdate(&PublishResponse{
+					Err: fmt.Errorf("could not get pending count for batch: %s: %w", batchID, err),
+				})
+				return nil
+			} else if count != 0 {
+				continue
+			}
+
+			jobs, err := n.repo.getByBatchID(ctx, batchID)
```

Comment:
I have a couple questions about this bit. I'm starting from the assumption that this code does the following in this order:
1. get the jobs and keep them in memory
2. delete the jobs, if something goes wrong return
3. send jobs to the consumer

So my questions are:
1. can you describe the workflow of what happens if we delete the jobs and then the service crashes?
2. could we possibly get and delete them in the same query? :thinking: 

No need to make changes here as well, just asking in order to understand.

Reply:
> could we possibly get and delete them in the same query? 🤔

Yes.

Reply:
> can you describe the workflow of what happens if we delete the jobs and then the service crashes?

Since it's a master which is tracking for a published job, in case it crashes, upon restart it will delete all the jobs and once the UploadJobs are picked up for processing they are again sent to the notifier for processing.